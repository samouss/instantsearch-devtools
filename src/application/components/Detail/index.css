.Detail {
  flex: 1;
  padding: 10px;
}

.DetailHeader {
  display: flex;
  align-items: center;
  top: 0;
  padding-top: 10px;
  padding-bottom: 10px;
  border-bottom: 1px solid lightgray;
  margin-top: -10px;
  margin-bottom: 10px;
  position: sticky;
  background-color: white;
  z-index: 999;
}

.DetailItemDate {
  font-weight: 600;
  margin-left: 10px;
}

.DetailSearchResults {
  margin-bottom: 30px;
}

.DetailSearchParameters {

}

.DetailSearchTitle {
  font-size: 1.2rem;
  margin-top: 0;
  margin-bottom: 10px;
}
