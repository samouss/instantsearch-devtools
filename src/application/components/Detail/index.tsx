import React, { SFC } from 'react';
import { Event } from '../../types';
import Type from '../Type';
import DateTime from '../DateTime';
import JSONViewer from '../JSONViewer';
import './index.css';

type Props = {
  event: Event;
};

const Detail: SFC<Props> = ({ event }) => (
  <div styleName="Detail">
    <div styleName="DetailHeader">
      <Type name={event.type} />
      <span styleName="DetailItemDate">
        <DateTime value={event.time} format="mm:ss.SSS" />
      </span>
    </div>

    {event.type === 'RESULT' && (
      <div styleName="DetailSearchResults">
        <pre styleName="DetailSearchTitle">searchResults</pre>
        <JSONViewer data={event.results} />
      </div>
    )}

    <div styleName="DetailSearchParameters">
      <pre styleName="DetailSearchTitle">searchParameters</pre>
      <JSONViewer data={event.parameters} />
    </div>
  </div>
);

export default Detail;
