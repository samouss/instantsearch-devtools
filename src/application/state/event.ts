import { v4 } from 'uuid';
import {
  HookEvent,
  ChangeHookEvent,
  SearchHookEvent,
  ResultHookEvent,
} from '../../types';
import {
  Id,
  State,
  Event,
  ChangeEvent,
  SearchEvent,
  ResultEvent,
} from '../types';

const START = Date.now();

const createChangeEvent = (event: ChangeHookEvent): ChangeEvent => ({
  type: 'CHANGE',
  id: v4(),
  time: Date.now() - START,
  parameters: event.parameters,
});

const createSearchEvent = (event: SearchHookEvent): SearchEvent => ({
  type: 'SEARCH',
  id: v4(),
  time: Date.now() - START,
  parameters: event.parameters,
});

const createResultEvent = (event: ResultHookEvent): ResultEvent => ({
  type: 'RESULT',
  id: v4(),
  time: Date.now() - START,
  parameters: event.parameters,
  results: event.results,
});

export const createEventFromHookEvent = (event: HookEvent): Event => {
  switch (event.type) {
    case 'CHANGE': {
      return createChangeEvent(event);
    }

    case 'SEARCH': {
      return createSearchEvent(event);
    }

    case 'RESULT': {
      return createResultEvent(event);
    }

    default: {
      throw new Error(`Event "${event.type}" is not supported.`);
    }
  }
};

const reducer = (state: State, event: Event): State => {
  switch (event.type) {
    case 'CHANGE': {
      return {
        ...state,
        eventIds: [event.id].concat(state.eventIds),
        eventById: state.eventById.set(event.id, event),
      };
    }

    case 'SEARCH': {
      return {
        ...state,
        eventIds: [event.id].concat(state.eventIds),
        eventById: state.eventById.set(event.id, event),
      };
    }

    case 'RESULT': {
      return {
        ...state,
        eventIds: [event.id].concat(state.eventIds),
        eventById: state.eventById.set(event.id, event),
      };
    }

    default: {
      return state;
    }
  }
};

export const getEvents = (state: State): Event[] => {
  return state.eventIds.map(id => {
    return state.eventById.get(id) as Event;
  });
};

export const getSelectedEvent = (state: State, id: Id): Event => {
  return state.eventById.get(id) as Event;
};

export default reducer;
