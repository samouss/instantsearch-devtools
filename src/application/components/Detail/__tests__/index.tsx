import React from 'react';
import { shallow } from 'enzyme';
import * as eventTestUtils from 'test/event';
import Detail from '../index';

describe('<Detail />', () => {
  it('expect to render with CHANGE event', () => {
    const props = {
      event: eventTestUtils.createFakeChangeEvent({
        parameters: {
          query: 'Apple',
        },
      }),
    };

    const wrapper = shallow(<Detail {...props} />);

    expect(wrapper).toMatchSnapshot();
  });

  it('expect to render with RESULT event', () => {
    const props = {
      event: eventTestUtils.createFakeResultEvent({
        parameters: {
          query: 'Apple',
        },
        results: {
          hits: [],
        },
      }),
    };

    const wrapper = shallow(<Detail {...props} />);

    expect(wrapper).toMatchSnapshot();
  });
});
