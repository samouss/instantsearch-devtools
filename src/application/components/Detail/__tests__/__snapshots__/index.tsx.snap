// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`<Detail /> expect to render with CHANGE event 1`] = `
<div
  styleName="Detail"
>
  <div
    styleName="DetailHeader"
  >
    <Type
      name="CHANGE"
    />
    <span
      styleName="DetailItemDate"
    >
      <DateTime
        format="mm:ss.SSS"
        value={1000}
      />
    </span>
  </div>
  <div
    styleName="DetailSearchParameters"
  >
    <pre
      styleName="DetailSearchTitle"
    >
      searchParameters
    </pre>
    <JSONViewer
      data={
        Object {
          "query": "Apple",
        }
      }
    />
  </div>
</div>
`;

exports[`<Detail /> expect to render with RESULT event 1`] = `
<div
  styleName="Detail"
>
  <div
    styleName="DetailHeader"
  >
    <Type
      name="RESULT"
    />
    <span
      styleName="DetailItemDate"
    >
      <DateTime
        format="mm:ss.SSS"
        value={3000}
      />
    </span>
  </div>
  <div
    styleName="DetailSearchResults"
  >
    <pre
      styleName="DetailSearchTitle"
    >
      searchResults
    </pre>
    <JSONViewer
      data={
        Object {
          "hits": Array [],
        }
      }
    />
  </div>
  <div
    styleName="DetailSearchParameters"
  >
    <pre
      styleName="DetailSearchTitle"
    >
      searchParameters
    </pre>
    <JSONViewer
      data={
        Object {
          "query": "Apple",
        }
      }
    />
  </div>
</div>
`;
